### PR TITLE
Issue/7857 generate product variants

### LINKED ISSUE #7857: Support batched "create variations" endpoint
### Description

We have to add support for the networking layer to request creating many variations with a single endpoint.

We'll use this endpoint from Woo REST API: https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-product-variations

### PR DESCRIPTION
**Pull request by atorresveiga**
Closes: https://github.com/woocommerce/woocommerce-android/issues/7857

### Why
We need to be able to create many variations for the Generate all combinations of variations project.
Creating all the variations in one request is the solution we agreed on after discussion, taking into account that it is fast and makes good use of resources. To achieve this solution, we'll use this endpoint from Woo REST API: https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-product-variations.

### Description
This PR adds support to create product variations using the batch-update-product-variations endpoint. It accomplishes its goal by modifying the batchUpdateVariations function and adding the missing parameters to match the API request. It also adds a UI to the FluxC example app to make it easier to test the request.

### Testing instructions

1. Open the FluxC example app
2. Tap on Woo
3. Tap on Select Site
4. Select a Site
5. Tap on Products
6. Tap on Select Site
7. Select a Site
8. Scroll to Batch Generate Variations 
9. Tap on Batch Generate Variations 
10. Enter a valid variable product id
11. Tap on Set Product ID
12. Enter a valid Attribute Id, Name, and Option
13. Tap on Add Attribute
14. Tap on Add Variation
15. Repeat steps 12 to 14 (as many variations as you want to add)
16. Tap on Generate Variation
17. Check that the variations are created successfully

### Images

| Adding variants | Success response |
| ------------- | ------------- |
|   |   |

### CODE REVIEW

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt [hunk 1]
```diff
@@ -932,34 +933,50 @@ class ProductRestClient @Inject constructor(
...
-        modifiedProperties: Map<String, Any>
-    ): WooPayload<BatchProductVariationsUpdateApiResponse> = WOOCOMMERCE.products.id(productId).variations.batch.pathV3
-        .let { url ->
-            val variationsUpdates: List<Map<String, Any>> = variationsIds.map { variationId ->
-                modifiedProperties.toMutableMap()
-                    .also { properties -> properties["id"] = variationId }
+        createVariations: List<Map<String, Any>>? = null,
+        updateVariations: List<Map<String, Any>>? = null,
+        deleteVariations: List<Long>? = null
+    ): WooPayload<BatchProductVariationsApiResponse> =
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt:956]: Added the missing parameters to the batch update request. The`createVariations` was needed to create the new variations (our goal), and the `deleteVariations` was added to match the API request. Now the rest client call looks like API request and can be used as documentation.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 2]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
         val variation: WCProductVariationModel
     ) : Payload<BaseNetworkError>()
 
+    class BatchGenerateVariationsPayload(
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt:176]: Named the Payload and the function `generate` instead of `create` with the intent to help future maintainers to realize that this function will only receive the `attributes` property. The rest of the variations properties will use their default values. The idea is to leave room for a future `create` function that will set the rest of the properties of the variations.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt [hunk 3]
```diff
@@ -6,7 +6,11 @@ import org.wordpress.android.fluxc.network.Response
...
  * Representation of Batch Update Product Variations API response.
  */
-class BatchProductVariationsUpdateApiResponse : Response {
+class BatchProductVariationsApiResponse : Response {
+    @SerializedName("create")
+    var createdVariations: List<ProductVariationApiResponse>? = null
     @SerializedName("update")
     var updatedVariations: List<ProductVariationApiResponse>? = null
+    @SerializedName("delete")
+    var deletedVariations: List<Long>? = null
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt]: I know you followed the pattern here but I think we could safely change those fields to `val` as they shouldn't be mutable.

I also think this class could be a `data` one as it's a data container.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 4]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
...
+        val site: SiteModel,
+        val remoteProductId: Long,
+        val variations: List<List<Map<String, Any>>>
+    ) : Payload<BaseNetworkError>() {
+        class Builder(
+            private val site: SiteModel,
+            private val remoteProductId: Long
+        ) {
+            private val variations = mutableListOf<List<Map<String, Any>>>()
+            fun getVariations() = variations.toList()
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: ❓ Why do we need this method?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 5]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
...
+
+            private val currentAttributes = mutableListOf<Map<String, Any>>()
+            fun getCurrentAttributes() = currentAttributes.toList()
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: ❓ Why do we need this method?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 6]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
         val variation: WCProductVariationModel
     ) : Payload<BaseNetworkError>()
 
+    class BatchGenerateVariationsPayload(
+        val site: SiteModel,
+        val remoteProductId: Long,
+        val variations: List<List<Map<String, Any>>>
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Can we create a new class to help understand this nested structure for API consumers? Without deep knowledge of variations, I can't know what parameters I should put - where's the attribute name and where's the term value e.g.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 7]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
...
+            private val currentAttributes = mutableListOf<Map<String, Any>>()
+            fun getCurrentAttributes() = currentAttributes.toList()
+
+            fun addVariationAttribute(id: Long, name: String, option: String) {
+                currentAttributes.add( buildMap {
+                    put("id", id)
+                    put("name", name)
+                    put("option", option)
+                })
+            }
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: This does not precisely add a variation attribute. "Attribute" is the whole thing: its name and all its options (terms).

In WooCommerce, the thing we add here is named "TermAssignment" (it's my proposal).

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 8]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
...
+
+            fun addVariation() {
+                if(currentAttributes.isEmpty()) return
+                variations.add(currentAttributes.toList())
+                currentAttributes.clear()
+            }
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Why do we need this method? I mean I understand what it does, but can't we simplify the `BatchGenerateVariationsPayload` by just passing received `variations`? This reduces overhead for library consumers.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt [hunk 9]
```diff
@@ -172,6 +172,43 @@ class WCProductStore @Inject constructor(
...
+            private val site: SiteModel,
+            private val remoteProductId: Long
+        ) {
+            private val variations = mutableListOf<List<Map<String, Any>>>()
+            fun getVariations() = variations.toList()
+
+            private val currentAttributes = mutableListOf<Map<String, Any>>()
+            fun getCurrentAttributes() = currentAttributes.toList()
+
+            fun addVariationAttribute(id: Long, name: String, option: String) {
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: ~⚙️ We shouldn't need to provide an `id` here. It sounds like a core responsibility - it'll be quite challenging to do this on the mobile side. Is it something that Core requests or can we skip it?~

Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Sorry! I'm wrong. It's `id` of attribute right? So we don't have to generate it, we have it from API. 

So if we have `id` of the `attribute` do we need the `name`? I think we shouldn't.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 7 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: > This does not precisely add a variation attribute. "Attribute" is the whole thing: its name and all its options (terms)

I believe this is true for the attribute property of a **Product**, which is different than the attribute property of a **Variation**:

#### Product ` /wc/v3/products/847/&_method=get`
```
{
 id: 2
 name: "Size"
▶options: ["S", "M", "L"]
 position: 0
 variation: true
 visible: true
}
```
-------------------
#### Variation  `/wc/v3/products/847/variations/&_method=get`
```
{
 id: 2
 name: "Size"
 option: "L"
}
```


> In WooCommerce, the thing we add here is named "TermAssignment" (it's my proposal).

Not sure about this. IMO using a different name than the one used by the API could be hard to follow, and I feel that the name `option` fits well with what we are trying to describe here. Actually, _**option**_ is the name used by other apps to represent this concept. That said, if you are strong about using **_term_** instead of **_option_**. I will be happy to change it.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt (hunk 3 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt]: I was hesitant about using data classes because when I checked the product API response classes, they all used this pattern of a class with vars and an empty constructor. However, I agree that a data class here is a better representation.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 9 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: >  Is it something that Core requests or can we skip it?

Yes, we need to provide the attribute `id` here; otherwise, the variation won't be added accordingly to our plan.

Let me walk you through an example using the data from the attributes of the product above.
In my store, I have a variable product with the attribute Size. The attribute property looks like this:

```
{
 id: 2
 name: "Size"
▶options: ["S", "M", "L"]
 position: 0
 variation: true
 visible: true
}
```
*Note that the attribute id is 2*

If I sent a request without the id, Core will create an empty variation like in the image above:



Bonus test, If I change the name the variation would be created with the right name:

Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Yes, thanks for the clarification. But can we then remove `name`?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 7 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: > I believe this is true for the attribute property of a Product, which is different than the attribute property of a Variation:

So you suggest:

- Leave `com.woocommerce.android.model.ProductAttribute` as it list: with `name: String` and `terms: List<String>`
- Introduce `com.woocommerce.android.model.VariationAttribute` which would look like:
```kotlin
data class VariationAttribute(
    val productAttributeName: String,
    val productAttributeId: Long,
    val selectedProductTerm: String
)
```
right? I'm okay with that although I'd consider if we also need `productAttributeName` if we have `productAttributeId` (this [comment](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569#discussion_r1025489174))

I'd keep naming the selection of attribute a "term" for reasons I've described [here](https://github.com/woocommerce/woocommerce-android/pull/7864#issuecomment-1319307375).

> Not sure about this.

I also didn't like `TermAssignment` 😅 .

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt (hunk 3 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/BatchProductVariationsApiResponse.kt]: Yeah, let's maybe leave it with the same pattern as rest of the API responses to not worry about it 👍

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 7 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: I just found there's `com.woocommerce.android.model.VariantOption`, so I think we should stick to that instead of a new `VariationAttribute`, right?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 8 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: > Why do we need this method? I mean I understand what it does, but can't we simplify the BatchGenerateVariationsPayload by just passing received variations?

The idea was to provide a Builder to help with the creation of the Payload class.  I used those methods on the example tests and thought it could be helpful when we plan to _build_ the payload (step by step) and not only pass the list of variations. The `getCurrentAttributes()` and `getCurrentAttributes()` were helper methods to check the payload builder status and see how many variations were in the payload builder and, how many attributes those variations have. 
One thing is the Builder a class responsible for helping build the Payload, which abstracts the building phase, and the other is the Payload, which do receives the variations in the constructor in the form of `val variations: List<List<Map<String, Any>>>`. I will take a look at the `VariationAttribute` that you recommended and refactor this class to make it more readable.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 5 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Answered [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569#discussion_r1026309567)

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 4 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Answered [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569#discussion_r1026309567)

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 9 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Thanks for the review @wzieba!🙏🏻 
I'll work on refactoring the Payload and the Builder class and let you know when the code is ready for another round.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 8 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Ok I see: I've used it as intended (I think) in this PR: https://github.com/woocommerce/woocommerce-android/pull/7893

If the consumer should use `Builder` can we make constructor of `BatchGenerateVariationsPayload` private (or internal) so there will be no doubt how to use it?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 7 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Added comment in different PR related to that: https://github.com/woocommerce/woocommerce-android/pull/7897#discussion_r1026319645

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 9 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: > But can we then remove name?

Is it possible to remove it? Yes, by the test, one can see that it is not used by Core at all. 
Still, I will vote to send the whole object because I don't have the context of why Core is not using it.
My concern is that they start using it in the future, and this breaks the app. This kind of change, breaking the API contract, will be hard to track in the future if you don't have the full context, and I don't see a huge impact on not sending the value. Not sure if I'm missing something here. WDYT?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 8 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: > The getCurrentAttributes() and getCurrentAttributes() were helper methods to check the payload builder status and see how many variations were in the payload builder and, how many attributes those variations have.

I see, thanks for clarification. Ideally, those methods should be `internal`, but I understand this is challenging because of `example` module being just a separate module.

Can we annotate those two methods by `org.jetbrains.annotations.ApiStatus.Internal`?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 9 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: I think we won't break API contract here; contrary, we'll satisfy it. Looking at the example in the [documentation](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569#discussion_r1026331922):
image

the example shows we send only `id` and `option` values. Not sending `name` seems reasonable because `id` always points to one concrete `name` - it's functional dependant.

It's not a huge deal, I agree. But if we want to follow the API, I think we should drop `name` (and reduce payload size as a bonus). WDYT?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 6 above)
Comment (wzieba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: As you explained here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569#discussion_r1026309567 I think we should rather make the constructor of the `BatchGenerateVariationsPayload` private.

If it'll be private, I think we don't need to provide new class - it's not exposed to consumers.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 7 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: I found `org.wordpress.android.fluxc.model.ProductVariantOption` and will stick to this instead of creating a new class. Here is the refactor using this class [b2770cc](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2569/commits/b2770cc083e764117715e8312501026ca5006796)

**Comment by atorresveiga** [comment]
Hi @wzieba! This PR is ready for another round 😄

**Review by wzieba — APPROVED** [review]
Looks good! Thank you for all the changes and discussions that brought improvements in both FluxC and Woo codebases regarding this feature 🙏 .

Let's check what's wrong with CI and merge this PR 👍

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 6 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Sorry, I missed these comments and refactored the class 😬

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt (hunk 9 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Good catch. I missed that in the example we send only `id` and `option` values. I'll create a new issue to improve this, as we discussed internally.

Comment (atorresveiga) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt]: Issue created here https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2570
